### ISSUE TITLE
Merge type stubs for torch.optim

### ISSUE THREAD

**Issue by ringohoffman** [pytorch/pytorch#102428]
### 🚀 The feature, motivation and pitch

Related #91648

@ezyang said that once `torch` dropped support for Python 2, `.pyi` files started being merged into `.py` files. As you can see in the linked example, having separate type stubs:

1. Masks methods the existence of methods, e.g. [`torch.optim.Optimizer.__getstate__`](https://github.com/pytorch/pytorch/blob/c9f4f01981fd73fcc7c27676cc50230cd1b5bc22/torch/optim/optimizer.pyi#L17), which can lead to spurious type errors in child classes of these classes.
2. Essentially turns off `mypy` for functions whose type hints are defined only in the stubs, since `mypy` doesn't run on functions/methods without inline type hints. This means we may have otherwise obvious type errors lurking in plain sight.

Overall, merging these type stubs should improve static type analysis in both `torch` and its consumers.

### Alternatives

Keep maintaining separate type stubs files.

cc @vincentqb @jbschlosser @albanD @janeyx99

**Comment by ezyang** [comment]
yes, will take patches for these. merging the stubs can be quite fiddly sometimes though

**Comment by janeyx99** [comment]
yep! happy to review PRs to do this!
